qiskit/providers/experiment/experiment_data.py

```diff
@@ -0,0 +1,674 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Experiment data class."""
+
+import logging
+import uuid
+import json
+from abc import abstractmethod
+from typing import Optional, List, Any, Union, Tuple, Callable, Dict
+import copy
+from concurrent import futures
+import queue
+from collections import OrderedDict, defaultdict
+from functools import wraps
+import traceback
+
+from qiskit.version import __qiskit_version__
+from qiskit.providers import Job, BaseJob, Backend, BaseBackend
+from qiskit.result import Result
+from qiskit.providers import JobStatus
+from qiskit.exceptions import QiskitError
+
+from .exceptions import ExperimentError, ExperimentEntryNotFound, ExperimentEntryExists
+from .analysis_result import AnalysisResultV1 as AnalysisResult
+from .json import NumpyEncoder, NumpyDecoder
+from .utils import save_data
+
+LOG = logging.getLogger(__name__)
+
+
+def auto_save(func: Callable):
+    """Decorate the input function."""
+    @wraps(func)
+    def _wrapped(self, *args, **kwargs):
+        return_val = func(self, *args, **kwargs)
+        if self.auto_save:
+            self.save()
+        return return_val
+    return _wrapped
+
+
+class ExperimentData:
+    """Base common type for all versioned ExperimentData classes.
+
+    Note this class should not be inherited from directly, it is intended
+    to be used for type checking. When implementing a provider you should use
+    the versioned abstract classes as the parent class and not this class
+    directly.
+    """
+    version = 0
+
+
+class ExperimentDataV1(ExperimentData):
+    """Class to handle experiment data.
+
+    This class serves as a data container for experiment data and any analysis
+    results or figures associated with the experiment. It also provides methods
+    used to interact with the database, such as saving the experiment data in
+    the database or retrieving its figured.
+    """
+
+    version = 1
+    _metadata_version = 1
+    _executor = futures.ThreadPoolExecutor()
+    """Threads used for asynchronous processing."""
+
+    _json_encoder = NumpyEncoder
+    _json_decoder = NumpyDecoder
+
+    def __init__(
+            self,
+            backend: Union[Backend, BaseBackend],
+            experiment_type: str,
+            experiment_id: Optional[str] = None,
+            tags: Optional[List[str]] = None,
+            job_ids: Optional[List[str]] = None,
+            share_level: Optional[str] = None,
+            metadata: Optional[Dict] = None,
+            figure_names: Optional[List[str]] = None,
+            notes: Optional[str] = None
+    ):
+        """Initializes the experiment data.
+
+        Args:
+            backend: Backend the experiment runs on. It can either be a
+                :class:`~qiskit.providers.Backend` instance or just backend name.
```

Comment:
The documentation here says that `backend` can be a backend name, however the type hints for the `backend` parameter do not include `str` type

Reply:
`str` (or backend name) will not work, because the code contains several instances of `self._backend.provider()` (for example in the `add_data` method)

Reply:
It should be possible to provide a `None` value to `backend`, for the case where the results come from somewhere and we want to continue from this point on

Reply:
fix in 42833a9